### DISCUSSION TITLE
Terminal States in GenerativeBeliefMDP

### DISCUSSION THREAD

**Discussion by FlyingWorkshop** [JuliaPOMDP/POMDPs.jl#555] [Ideas]
Porting this conversation from email @zsunberg, @mykelk.

Is there another way to implement [@gen in GenerativeBeliefMDP](https://github.com/JuliaPOMDP/POMDPModelTools.jl/blob/ba3ab10864bdbefd0e3d4bee6aa0ba523c935877/src/generative_belief_mdp.jl#L17-L26). Currently, the function samples a state from the belief and checks whether the sampled state is terminal _during generation_. This makes it hard to check and handle termination _before generation_ (which is a useful behavior in practice).

**Comment by zsunberg** [comment]
At one level higher, what are you trying to do? i.e. why is checking and handling termination before generation "a useful behavior in practice"? or what would you do if you could check it before next state generation? (maybe link to some code where there is a problem)

**Reply by FlyingWorkshop** [reply]
This is useful when you want to rollout trajectories and stop at a terminal state (see [snippet](https://github.com/JuliaPOMDP/CompressedBeliefMDPs.jl/blob/0672139016e748953bd21f566f9063dad68eb5c7/src/samplers/rollout.jl#L57-L61)).

**Reply by zsunberg** [reply]
I see. I will think about a permanent fix for `GenerativeBeliefMDP`

For now, to get your code to run, I'd recommend this:
```julia
function (s::PolicySampler)(pomdp::POMDP)
    return unique(stepthrough(pomdp, s.policy, s.updater, "bp", max_steps=s.n))
end
```
or similar. I think that will do what you want.

**Reply by zsunberg** [reply]
The behavior was improved in #559

**Comment by zsunberg** [comment]
I spoke to my PhD students who had a similar issue, and they decided to use problems where the observation indicates that the state must be terminal so that the belief is always terminal when the state is terminal.
